### PR TITLE
Remove extra type error after missing semicolon error

### LINKED ISSUE #103425: Strange extra type error after missing semicolon error
Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=55059b7460a49a908e086ce00bf462e2)):

```rust
fn f() -> f32 {
    3
    5.0
}

fn main() {
}
```

The current output is:

```
error: expected `;`, found `5.0`
 --> epp.rs:2:6
  |
2 |     3
  |      ^ help: add `;` here
3 |     5.0
  |     --- unexpected token

error[E0308]: mismatched types
 --> epp.rs:2:5
  |
2 |     3
  |     ^ expected `()`, found integer
```

What is this second error trying to tell me? _What_ did it expect to be unit type and why? Does it make sense for this error to be emitted despite the first error correctly identifying my problem?

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #103425

### CODE REVIEW

File: compiler/rustc_parse/src/parser/stmt.rs [hunk 1]
```diff
@@ -555,9 +555,15 @@ impl<'a> Parser<'a> {
             StmtKind::Expr(ref mut expr)
                 if self.token != token::Eof && classify::expr_requires_semi_to_be_stmt(expr) =>
             {
+
+                let res = self.expect_one_of(&[], &[token::Semi, token::CloseDelim(Delimiter::Brace)]);
+                let mut skip_type_error = false;
+                if let Ok(true) = &res {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/stmt.rs:561]: Please use a match here instead of assigning `res` to a local variable.

Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/stmt.rs:561]: Also, can you comment _why_ this works? It's not clear from the usage that `Ok(true)` means "recovered".

File: compiler/rustc_parse/src/parser/stmt.rs [hunk 2]
```diff
@@ -585,7 +591,11 @@ impl<'a> Parser<'a> {
                         e.emit();
                         self.recover_stmt();
                     }
-                    // Don't complain about type errors in body tail after parse error (#57383).
+                    skip_type_error = true;
+                }
+
+                if skip_type_error {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/stmt.rs:597]: Can you give this a different name? `skip_type_error` -> `replace_with_err` or something?

File: src/test/ui/parser/issue-103425.rs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+fn f() -> f32 {
+    3
+    //~^ ERROR expected `;`
+    5.0
+}
```
Comment (compiler-errors) [inline src/test/ui/parser/issue-103425.rs:5]: Do you mind adding the other test where we have several missing semicolons? https://github.com/rust-lang/rust/issues/103425#issuecomment-1288121143

```
fn f() -> f32 {
    2_u32
    3_i8
    5.0
}

fn main() {
}
```

Comment (chenyukang) [reply src/test/ui/parser/issue-103425.rs:5]: Added it.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**